Error `Option skip is only valid for smtbmc and btor engines` running riscv-formal

I'm trying to add riscv-formal to my core and while running it, I'm getting:

```sh
❯ ./start_formal.sh
Reading checks.cfg.
Creating checks directory.
Generated 43 checks.
make: Entering directory '/src/cores/chiselv/checks'
sby causal_ch0.sby
sby liveness_ch0.sby
sby pc_bwd_ch0.sby
sby pc_fwd_ch0.sby
sby reg_ch0.sby
sby unique_ch0.sby
sby insn_add_ch0.sby
sby insn_addi_ch0.sby
sby insn_and_ch0.sby
sby insn_andi_ch0.sby
sby insn_auipc_ch0.sby
sby insn_beq_ch0.sby
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [pc_bwd_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
SBY 13:30:01 [liveness_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
make: *** [makefile:12: pc_bwd_ch0/status] Error 1
make: *** Waiting for unfinished jobs....
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [unique_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
make: *** [makefile:8: liveness_ch0/status] Error 1
make: *** [makefile:24: unique_ch0/status] Error 1
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [reg_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
make: *** [makefile:20: reg_ch0/status] Error 1
SBY 13:30:01 [insn_addi_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [causal_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [pc_fwd_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
make: *** [makefile:32: insn_addi_ch0/status] Error 1
SBY 13:30:01 [insn_and_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [insn_andi_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
SBY 13:30:01 [insn_auipc_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
make: *** [makefile:4: causal_ch0/status] Error 1
make: *** [makefile:16: pc_fwd_ch0/status] Error 1
make: *** [makefile:40: insn_andi_ch0/status] Error 1
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
SBY 13:30:01 [insn_add_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
WARNING: Could not connect to jobserver specified in MAKEFLAGS, disabling parallel execution.
make: *** [makefile:36: insn_and_ch0/status] Error 1
SBY 13:30:01 [insn_beq_ch0] ERROR: Option skip is only valid for smtbmc and btor engines.
make: *** [makefile:28: insn_add_ch0/status] Error 1
make: *** [makefile:44: insn_auipc_ch0/status] Error 1
make: *** [makefile:48: insn_beq_ch0/status] Error 1
make: Leaving directory '/src/cores/chiselv/checks'
```

My config is:

```cfg
[options]
isa rv32i

[depth]
insn            20
reg       15    30
pc_fwd    10    30
pc_bwd    10    30
liveness  1  10 30
unique    1  10 30
causal    10    30

[defines]
`define RISCV_FORMAL_ALIGNED_MEM

[script-sources]
read_verilog -sv @basedir@/cores/@core@/wrapper.sv
read_verilog @basedir@/cores/@core@/RVFITop.v
```

This also happens to existing cores like vexriscv or picorv32.

I'm running yosys and sby from `gcr.io/hdl-containers/formal` container which uses:

```
root@6f4ae94e1682:/src/cores/chiselv# yosys --version
Yosys 0.20+74 (git sha1 9313549cd, clang 11.0.1-2 -fPIC -Os)
```

Latest SBY from https://github.com/YosysHQ/sby.

--- a/sbysrc/sby_core.py
+++ b/sbysrc/sby_core.py
@@ -305,7 +305,9 @@ class SbyConfig:
                 if section == "engines":
                     mode = "engines"
 
-                    if args is not None:
+                    if args is None:
+                        engine_mode = None
+                    else:
                         section_args = args.split()
 
                         if len(section_args) > 1:
@@ -314,11 +316,16 @@ class SbyConfig:
                         if section_args[0] not in ("bmc", "prove", "cover", "live"):
                             self.error(f"sby file syntax error: Expected one of 'bmc', 'prove', 'cover', 'live' as '[engines]' argument, got '{section_args[0]}'")
 
-                        if section_args[0] in self.engines:
-                            self.error(f"Already defined engine block for mode '{section_args[0]}'")
+                        engine_mode = section_args[0]
+
+                    if engine_mode in self.engines:
+                        if engine_mode is None:
+                            self.error(f"Already defined engine block")
                         else:
-                            self.engines[section_args[0]] = list()
-                            engine_mode = section_args[0]
+                            self.error(f"Already defined engine block for mode '{engine_mode}'")
+                    else:
+                        self.engines[engine_mode] = list()
+
                     continue
 
                 if section == "setup":
@@ -415,8 +422,6 @@ class SbyConfig:
 
             if mode == "engines":
                 args = line.strip().split()
-                if engine_mode not in self.engines:
-                    self.engines[engine_mode] = list()
                 self.engines[engine_mode].append(args)
                 continue
 
@@ -642,7 +647,8 @@ class SbyTask(SbyConfig):
                     print(line, file=f)
 
     def engine_list(self):
-        return list(enumerate(self.engines.items()))
+        engines = self.engines.get(None, []) + self.engines.get(self.opt_mode, [])
+        return list(enumerate(engines))
 
     def check_timeout(self):
         if self.opt_timeout is not None:

--- a/sbysrc/sby_mode_bmc.py
+++ b/sbysrc/sby_mode_bmc.py
@@ -24,20 +24,8 @@ def run(task):
     task.handle_int_option("append", 0)
     task.handle_str_option("aigsmt", "yices")
 
-    for engine_idx, engine_section in task.engine_list():
-        if isinstance(engine_section, list):
-            engine = engine_section
-            engine_name = None
-        else:
-            assert len(engine_section[1]) > 0
-            engine = engine_section[1][0]
-            engine_name = engine_section[0]
-
-        if engine_name is None:
-            engine_name = engine_idx
-
-
-        task.log(f"""engine_{engine_name}: {" ".join(engine)}""")
+    for engine_idx, engine in task.engine_list():
+        task.log(f"""engine_{engine_idx}: {" ".join(engine)}""")
         task.makedirs(f"{task.workdir}/engine_{engine_idx}")
 
         if engine[0] == "smtbmc":

--- a/sbysrc/sby_mode_cover.py
+++ b/sbysrc/sby_mode_cover.py
@@ -23,19 +23,8 @@ def run(task):
     task.handle_int_option("depth", 20)
     task.handle_int_option("append", 0)
 
-    for engine_idx, engine_section in task.engine_list():
-        if isinstance(engine_section, list):
-            engine = engine_section
-            engine_name = None
-        else:
-            assert len(engine_section[1]) > 0
-            engine = engine_section[1][0]
-            engine_name = engine_section[0]
-
-        if engine_name is None:
-            engine_name = engine_idx
-
-        task.log(f"""engine_{engine_name}: {" ".join(engine)}""")
+    for engine_idx, engine in task.engine_list():
+        task.log(f"""engine_{engine_idx}: {" ".join(engine)}""")
         task.makedirs(f"{task.workdir}/engine_{engine_idx}")
 
         if engine[0] == "smtbmc":

--- a/sbysrc/sby_mode_live.py
+++ b/sbysrc/sby_mode_live.py
@@ -24,19 +24,8 @@ def run(task):
 
     task.status = "UNKNOWN"
 
-    for engine_idx, engine_section in task.engine_list():
-        if isinstance(engine_section, list):
-            engine = engine_section
-            engine_name = None
-        else:
-            assert len(engine_section[1]) > 0
-            engine = engine_section[1][0]
-            engine_name = engine_section[0]
-
-        if engine_name is None:
-            engine_name = engine_idx
-
-        task.log(f"""engine_{engine_name}: {" ".join(engine)}""")
+    for engine_idx, engine in task.engine_list():
+        task.log(f"""engine_{engine_idx}: {" ".join(engine)}""")
         task.makedirs(f"{task.workdir}/engine_{engine_idx}")
 
         if engine[0] == "aiger":

--- a/sbysrc/sby_mode_prove.py
+++ b/sbysrc/sby_mode_prove.py
@@ -31,19 +31,8 @@ def run(task):
     task.basecase_procs = list()
     task.induction_procs = list()
 
-    for engine_idx, engine_section in task.engine_list():
-        if isinstance(engine_section, list):
-            engine = engine_section
-            engine_name = None
-        else:
-            assert len(engine_section[1]) > 0
-            engine = engine_section[1][0]
-            engine_name = engine_section[0]
-
-        if engine_name is None:
-            engine_name = engine_idx
-
-        task.log(f"""engine_{engine_name}: {" ".join(engine)}""")
+    for engine_idx, engine in task.engine_list():
+        task.log(f"""engine_{engine_idx}: {" ".join(engine)}""")
         task.makedirs(f"{task.workdir}/engine_{engine_idx}")
 
         if engine[0] == "smtbmc":

--- a/tests/regression/option_skip.sby
+++ b/tests/regression/option_skip.sby
@@ -1,0 +1,33 @@
+[tasks]
+smtbmc_pass: smtbmc pass
+smtbmc_fail: smtbmc fail
+btormc_pass: btormc pass
+btormc_fail: btormc fail
+
+[options]
+mode bmc
+pass: expect pass
+fail: expect fail
+pass: depth 5
+fail: depth 6
+
+skip 2
+
+[engines]
+smtbmc: smtbmc boolector
+[engines bmc]
+btormc: btor btormc
+
+[script]
+read -formal top.sv
+prep -top top
+
+[file top.sv]
+module top(input clk);
+    reg [7:0] counter = 0;
+
+    always @(posedge clk) begin
+        counter <= counter + 1;
+        assert (counter < 4);
+    end
+endmodule

